OpenRA.Mods.Common/Activities/Rearm.cs

```diff
@@ -8,53 +8,65 @@
  */
 #endregion
 
+using System.Collections.Generic;
 using System.Linq;
 using OpenRA.Activities;
 using OpenRA.Mods.Common.Traits;
-using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Activities
 {
 	public class Rearm : Activity
 	{
-		readonly LimitedAmmo limitedAmmo;
-		int ticksPerPip = 25 * 2;
-		int remainingTicks = 25 * 2;
-		string sound = null;
+		readonly IEnumerable<AmmoPool> ammoPools;
+		readonly Dictionary<AmmoPool, int> ammoPoolsReloadTimes;
 
-		public Rearm(Actor self, string sound = null)
+		public Rearm(Actor self)
 		{
-			limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			if (limitedAmmo != null)
-				ticksPerPip = limitedAmmo.ReloadTimePerAmmo();
-			remainingTicks = ticksPerPip;
-			this.sound = sound;
+			ammoPools = self.TraitsImplementing<AmmoPool>();
+
+			if (ammoPools == null)
+				return;
+
+			ammoPoolsReloadTimes = ammoPools.ToDictionary(x => x, y => y.Info.ReloadTicks);
 		}
 
 		public override Activity Tick(Actor self)
 		{
-			if (IsCanceled || limitedAmmo == null)
+			if (IsCanceled || ammoPools == null)
```

Comment:
Rearming structures should only rearm ammo pools that don't self reload, right?
Or should they also speed up rearming of ammo pools that self reload?